app/models/user.rb

```diff
@@ -18,20 +44,59 @@ class User < ApplicationRecord
   has_many :chatrooms, through: :gigs
   has_one_attached :profile_pic
 
-  def self.from_omniauth(auth)
-    where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
-      user.email = auth.info.email
-      user.password = Devise.friendly_token[0, 20]
-      user.spotify_token = auth.credentials.token
-      user.spotify_refresh_token = auth.credentials.refresh_token
-      user.spotify_expires_at = Time.at(auth.credentials.expires_at)
+  def save_new_access_token(new_access_token)
+    self.update(spotify_access_token: new_access_token)
+    self.save
+  end
+
+  def spotify_accessible?
+    return false if self.spotify_access_token.nil?
+    return false if self.spotify_refresh_token.nil?
+
+    true
+  end
+
+  # https://github.com/guilhermesad/rspotify
+  def spotify_user
+    RSpotify::User.new(
+      {
+        "credentials" => {
+          "token" => self.spotify_access_token,
+          "refresh_token" => self.spotify_refresh_token,
+          "access_refresh_callback" =>
+            Proc.new { |new_access_token, token_lifetime |
+              now = Time.now.utc.to_i
+              deadline = now+token_lifetime
+              self.save_new_access_token(new_access_token)
+            }
+        },
+        "id" => self.spotify_id
+      }
+    )
+  end
+
+  # https://developer.spotify.com/documentation/web-api/reference/get-users-top-artists-and-tracks
+  def top_artists
+    return [] unless spotify_accessible? && spotify_user.present?
+
+    begin
+      spotify_user.top_artists
+    rescue
+      []
     end
   end
 
-  # WIP: Spotify API integration
-  def spotify_client
-    @spotify_client ||= RSpotify::User.new("credentials" => {
-                                             "token" => spotify_token, "refresh_token" => spotify_refresh_token, "expires_at" => spotify_expires_at.to_i,
-                                           })
+  # https://developer.spotify.com/documentation/web-api/reference/get-users-top-artists-and-tracks
+  def top_tracks
+    return [] unless spotify_accessible?
+
+    spotify_user.top_tracks
+  end
+
+  # https://developer.spotify.com/documentation/web-api/reference/get-followed
+  def following_artists
+    return [] unless spotify_accessible?
+
+    spotify_user.following(type: "artist")
```

Comment:
These are new wrapper methods for spotify. One reason to _not_ merge this before demo is that these can potentially fail. They are not fully tested and things like users rejecting auth from Spotify's UI may cause oddities. Thus this section would need greater testing even post demo day.